std/range/package.d

```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+            _nextSource = source.save;
+            _nextSource.popFrontN(windowSize);
+        }
+
+        static if (!slidesWithLessElements)
+        {
+            // empty source range is needed, s.t. length, slicing etc. works properly
+            static if (needsEndTracker)
+            {
+                if (_nextSource.empty)
+                {
+                    _source = _nextSource;
+                }
+            }
+            else
+            {
+                if (_source.length < windowSize)
+                {
+                    static if (hasSlicing!Source)
+                    {
+                        static if (hasSliceToEnd)
+                            _source = _source[$ .. $];
+                        else
+                            _source = _source[_source.length .. _source.length];
+                    }
+                    else
+                    {
+                        _source.popFrontN(_source.length);
+                    }
+                }
+            }
+        }
+
+        _empty = _source.empty;
+    }
+
+    /// Forward range primitives. Always present.
+    @property auto front()
+    {
+        assert(!empty, "Attempting to access front on an empty range");
+        static if (hasSlicing!Source && hasLength!Source)
+        {
+            immutable len = _source.length;
+            immutable end = (len > _windowSize) ? _windowSize : len;
+            return _source[0 .. end];
+        }
+        else
+        {
+            return _source.save.take(_windowSize);
+        }
+    }
+
+    /// Ditto
+    void popFront()
+    {
+        assert(!empty, "Attempting to call popFront() on an empty range");
+        _source.popFrontN(_stepSize);
+
+        static if (needsEndTracker)
+        {
+            if (_nextSource.empty)
+                _empty = true;
+            else
+                _nextSource.popFrontN(_stepSize);
+        }
+        else
+        {
+            if (_source.length < _windowSize)
+                _empty = true;
+        }
+    }
+
+    static if (!isInfinite!Source)
+    {
+        /// Ditto
+        @property bool empty()
+        {
+            return _empty;
+        }
+    }
+    else
+    {
+        // undocumented
+        enum empty = false;
+    }
+
+    /// Ditto
+    @property typeof(this) save()
+    {
+        return typeof(this)(_source.save, _windowSize, _stepSize);
+    }
+
+    static if (hasLength!Source)
+    {
+        /// Length. Only if $(D hasLength!Source) is $(D true)
+        @property size_t length()
+        {
+            if (_windowSize > _source.length)
+            {
+                static if (slidesWithLessElements)
+                    return 1;
+                else
+                    return 0;
+            }
+            else
+            {
+                import std.math : ceil;
+                double t = _source.length - _windowSize + 1;
+                return cast(size_t) ceil(t / _stepSize);
```

Comment:
nah, that's overkill - use `(_source.length - _windowSize + _stepSize) / _stepSize`